### ISSUE TITLE
FTMTree weird segmentation

### ISSUE THREAD

**Issue by dsakurai** [topology-tool-kit/ttk#626]
**Describe the bug**

The segmentation generated from FTM contour tree is wrong (as far as I understand).

<img width="2091" alt="debug-contourtree-latest-commit" src="https://user-images.githubusercontent.com/16434308/122849277-29f49b80-d346-11eb-9c17-808de069c019.png">

In the image above, the segmentation is color-coded on the point gaussian of the mesh.
(Edit: the numbers on the vertices are the vertex IDs.)
The gray cylinders are the skeleton arcs of the contour tree.

Point 9 shares the segmentation with point 10.
This is a bug (if my understanding is correct) because the yellow contour containing point 4 should separate point 9 from 10.

I dug in the code to find out that the FTM tree somehow adds point 9 in the superarc between points 4 and 10.
I'm not perfectly sure about this observation because I don't fully understand the code. Please have a look if needed.

In addition, point 11 should also not belong to the same segmentation as point 10.

**To Reproduce**

Run the ParaView state file below:
[debug-contourtree.pvsm.zip](https://github.com/topology-tool-kit/ttk/files/6690438/debug-contourtree.pvsm.zip)
I used ParaView 5.9.0 with the latest dev branch of the TTK (commit 3fbb9e98fc2d9f4e80a8b79e09262b1cc74c3345 ).

**Expected behavior**
Point 9 and 11 should belong to a segmentation different from point 10.

**System:**
 - OS: macOS 10.15.7 Catalina
 - Compiler: clang 10
 - Installed TTK's optional dependencies:
~~~
 TTK_ENABLE_EIGEN                 ON                                                                                                                                                                    
 TTK_ENABLE_GRAPHVIZ              ON                                                                                                                                                                    
 TTK_ENABLE_OPENMP                ON                                                                                                                                                                    
 TTK_ENABLE_SCIKIT_LEARN          ON                                                                                                                                                                    
 TTK_ENABLE_SQLITE3               ON                                                                                                                                                                    
 TTK_ENABLE_ZLIB                  ON                                                                                                                                                                    

 TTK_ENABLE_EMBREE                OFF                                                                                                                                                                   
 TTK_ENABLE_MPI                   OFF                                                                                                                                                                   
 TTK_ENABLE_SPECTRA               OFF                                                                                                                                                                   
 TTK_ENABLE_WEBSOCKETPP           OFF                                                                                                                                                                   
 TTK_ENABLE_ZFP                   OFF                                                                                                                                                                   
~~~

**Comment by julien-tierny** [comment]
hi daisuke, 
thanks for your issue.
I have the impression that the values you attach on the vertices do not match with the color map (for instance, based on the color map, vertex 4 seems to be indeed higher than vertex 9).
I'll play with your data and come back to you shortly.
best,

**Comment by julien-tierny** [comment]
I confirm the issue. 
in the join tree, the vertex 9 is correctly located (below 4). the split tree segmentation looks OK.
it looks like a bug when combining the two into the contour tree.
@CharlesGueunet  any thoughts?

**Comment by CharlesGueunet** [comment]
Hello, I have had a quick look and I confirm there is a problem here. I won't have the time to look further until next week, but I will definitely have a look then.

Best

**Comment by julien-tierny** [comment]
awesome, thanks!
-- 
Dr Julien Tierny
CNRS Researcher
Sorbonne Universite
http://lip6.fr/Julien.Tierny

**Comment by dsakurai** [comment]
@julien-tierny, @CharlesGueunet

Thank you for your quick reply :-)
I'll wait for updates!

> the values you attach on the vertices do not match with the color map

My bad. The numbers on the vertices are the vertex IDs. I should have stated that.
Seems you guys figured that out, though.

**Comment by dsakurai** [comment]
Hi @CharlesGueunet,

I'm not in a hurry, but have you found the time to have a look? :-)

**Comment by CharlesGueunet** [comment]
I am looking at the issue, it seems it has been introduced by f6f21b0206cdbac6ba802d5afc5da31debae43ca, I will try to fix this as soon as I can

**Comment by CharlesGueunet** [comment]
not sure about the commit, but there was a mechanism used to sort scalars in reverse order for the join and split tree segmentation that I do not find anymore in the code, so when we insert nodes of the join tree in the split before the combination, the segmentation become wrong

**Comment by CharlesGueunet** [comment]
(FYI FTR gives the right result)

**Comment by CharlesGueunet** [comment]
The mentionned commit changed a lot of `<=` to `<` and this lead to wrong segmentation

**Comment by dsakurai** [comment]
Thank you :-)

**Comment by julien-tierny** [comment]
fixed with PR #634
